wallet: Prepare for future upgrades by recording versions of last client to open and decrypt

achow101 opened the pull request:
When a wallet is automatically upgraded, we always do so in a way that allows the user to load their wallet in an older version. When the user loads their wallet into an upgraded version again, the wallet may not perform the automatic upgrade and end up with upgraded and non-upgraded material in the wallet.

If we write some information about the last client to open the wallet, we would be able to detect these upgrade-downgrade-upgrade situations and perform another upgrade if so. However, in order for this to work, we need to have implemented writing such information into versions prior to the ones which introduce new automatic upgrades.

We are already writing the CLIENT_VERSION into all wallets in a "version" record, although this record is not updated in the same way across all previous versions. Since v24.0, we are always updating the record when the client version differs, but prior to that, the record would only be updated when the client version is newer. Even so, we can use this record to store the last client version.

The first commit decouples the written client version from the node version. This aids in development of new features. The version is entirely decoupled from previous versions by requiring all versions to have bit 19 set, and the initial version being `0 | (1 << 19)`. The versions just need to be at least 299900, and forcing bit 19 to be set was an easy way to achieve this.

However, using version numbers is not as flexible as feature flags, so the second commit introduces the concept of Wallet Client Features. These flags are distinct from the existing Wallet Feature Flags since they are conceptually different. These are written in a separate record, and the wallet client version is incremented to `(1 << 19) + 1`.

Lastly, since some upgrades may need private keys and can only occur after the wallet is decrypted, the third commit also adds a last decrypted features flags record to record the features of the last client to decrypt the wallet.

The result is that we will now be able to detect a downgrade down to v24.0 and be able to implement in parallel multiple features that require automatic upgrades.

achow101 committed:
walletdb: Decouple last client record from CLIENT_VERSION

Instead of tying the last client record with the node version, introduce
a separate wallet client version that will be increased as necessary
when new automatic upgrades are introduced.

achow101 committed:
wallet: Record the supported features of the last client to decrypt a wallet

ryanofsky commented:
Concept ACK. It shouldn't hurt to add more metadata to help detect when there might be a mix of old data in a newer format and new data in an older format in the same wallet.

Conceptually, I think we could always detect these cases by just scanning the entire wallet for data that needs to be upgraded on loading. But having this extra metadata is nice because it can avoid slowing down wallet loading by avoiding unnecessary upgrade rescans.

On the approach in e94ef51837daa5ec9552eb14e8ccc38d1b82b65b, I think it is ok, but would suggest some ways it might be improved:

- I agree with Marco's [comment](https://github.com/bitcoin/bitcoin/pull/32895#issuecomment-3048078577) that it would be great to move away from CLIENT_VERSION and use something independent of the node version.
  - A downside of this suggestion is that the PR might not be able to reuse the existing VERSION field and might need to introduce 2 new fields instead of 1. I think the tradeoff could be worth it though.
- It would be nice to avoid version numbers altogether and use feature flags instead, defining new feature flags when new data fields are introduced, and writing supported feature flags to LAST_OPENED_FEATURES & LAST_DECRYPTED_FEATURES fields on wallet loading.
  - This would not require introducing a new data type since we already have a [WalletFlags](https://github.com/bitcoin/bitcoin/blob/927055e42afbfd998b5d85108e9c5fe7c6ffb7aa/src/wallet/walletutil.h#L36-L78) enum. New internal and external features could just be new (nonmandatory) `WalletFlags` values.
  - IMO, flags are better than version numbers because they decouple code that implements new features from the merge and release process. They also make code more readable and maintainable because checks like `if (last_opened_supported_specific_feature_x)` are more obvious than `if (last_opened_version > some_version_number_with_collection_of_features)`
- I assume we don't care about detecting upgrades & downgrades to versions before v24.0 which wouldn't overwrite the `VERSION` field on opening. But if we did, it'd be possible to detect this by writing last_opened and last_decrypted metadata to a file alongside the wallet database instead of inside the wallet database, and touching the file whenever the database is flushed. This way, if the database file modification time was greater than metadata file modification time, you could infer that the database had been loaded by an older version in the meantime and needs to be upgraded. Wouldn't recommend this approach because it is more complicated, but wanted to mention it as an alternative to introducing new database fields.

achow101 commented:
> I wonder if this can be uncoupled from CLIENT_VERSION, because the coupling is confusing and makes testing harder:

Definitely, however I'd like to be able to reuse the existing "version" record which mostly has the desired behavior since v24.0. The detection only works because of older versions overwriting the field with their CLIENT_VERSION. Introducing a new field now means that we cannot detect when an upgraded wallet was opened in any version older than v30.0 (assuming this is merged for the next release). The timeline would be too soon for any newly introduced features in say v31.0 or v32.0 (and I've got a few things in mind already that would utilize this). I'm already not particularly thrilled by the fact that the current implementation can only detect down to 24.0 as I expect that users doing upgrade-downgrade-upgrade with even older versions is commonplace.

> Conceptually, I think we could always detect these cases by just scanning the entire wallet for data that needs to be upgraded on loading.

I don't think that is actually viable as it would largely end up performing the upgrade on all loads regardless of whether the upgrade has already been done.

achow101 commented:
> Maybe I am missing something obvious, but my suggestion was not to introduce a new field/record. It was to re-use the existing field (and value) and manually handle the value going forward only when needed.

Sorry, I was responding to @ryanofsky's comment at the same time since they were very similar.

I think that may be workable.

achow101 committed:
wallet: Introduce LastClientFeatures flags and LAST_OPENED_FEATURES record

LastClientFeatures are feature flags indicating what automatic upgrade
features supported by the last client that opened the wallet.

The LAST_OPENED_FEATURES record stores these flags and must be
set to match the exact features supported by the client that opens a
wallet.

achow101 commented:
I've updated the PR to do a combination of @maflcko's and @ryanofsky's suggestions.

The new approach decouples `version` from CLIENT_VERSION. It also writes out a set of client specific feature flags into two new records: `lastopenedfeatures` and `lastdecryptedfeatures`. However, these feature flags are separate from the existing wallet feature flags.

I believe this preserves the behavior of detecting downgrades down to v24.0, while also giving us the flexibility of a new field and feature flags.

achow101 committed:
wallet: Set last opened and decrypted features during migration

Set these flags to avoid the automatic upgrade after migrating.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32895/commits/b0470f573735a04ebc87ca5d3e00a73e3f10fee1

nit: Commit https://github.com/bitcoin/bitcoin/pull/32895/commits/3662960470d74850008b7b48b6820d02002f7a5d appears to have CI failures.

achow101 commented:
> nit: Commit [3662960](https://github.com/bitcoin/bitcoin/commit/3662960470d74850008b7b48b6820d02002f7a5d) appears to have CI failures.

I believe that's only because the tasks were canceled by a new push.

maflcko reviewed (commented):
(just the llm nits)

Bortlesboat reviewed (commented):
Concept ACK 67417e946e1dcf3e5bd8b22be06ad9bba3859efe

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -631,6 +631,16 @@ bool CWallet::Unlock(const SecureString& strWalletPassphrase)
+                if (!m_last_decrypted_features || *m_last_decrypted_features != WALLET_CLIENT_FEATURES) {
+                    // Write the current wallet client features to LAST_DECRYPTED_FEATURES.
+                    // This must be done after all automatic upgrades so that those upgrades can be
+                    // performed in an upgrade-downgrade-upgrade scenario.
```
It doesn't matter. The automatic upgrades are no-ops if they've already been done. If the write fails, the next time the wallet is loaded and unlocked, the upgrade gets done again, but that will do nothing. Really this is an optimization to avoid doing the work multiple times, but it doesn't functionally matter if it does happen multiple times.

ajtowns commented on src/wallet/walletutil.h:
```diff
@@ -56,6 +56,18 @@ enum WalletFlags : uint64_t {
+// New version numbers must be greater than 299900 which is guaranteed by setting bit 30
+enum WalletClientVersion : int32_t {
+    // OR with all versions to set bit 30
+    VERSION_MASK = (1ULL << 30),
```
`VERSION_LATEST = (1L << 19) + 0` would be sufficient here; seems something of a shame to waste 3/4ths of the available space (half due to negative values, quarter by going straight to the highest remaining bit). Using `1ULL` for a value going into an `int32_t` seems odd.

achow101 replied:
Changed `VERSION_MASK` to `MIN_VERSION = (1L << 19)`, and for the subsequent version number to increment on that.

ajtowns commented on src/wallet/walletutil.h:
```diff
@@ -65,7 +65,18 @@ enum WalletClientVersion : int32_t {
+    VERSION_LATEST = VERSION_LAST_CLIENT_FEATURES
+};
+
+enum LastClientFeatures : uint64_t {
```
What happens when you want to implement a 65th wallet feature?

Does `VERSION_LATEST` get bumped every time there's a new feature defined, or is that unnecessary? It would seem possible to only bump `VERSION_LATEST` when there's an incompatible change to features (some features change their meanings, the field size is increased etc). If so, it might be good for there to be some compatibility overlap, so that you can cleanly switch between software with wallet client version `x` and `x-1`, even if switching between `x` and `x-2` involves full rescans of wallet features. (With 64 feature bits, bumping from `x` to `x+1` is probably at least five or six years of releases before bits need to be reused)

achow101 replied:
> What happens when you want to implement a 65th wallet feature?

I think what would happen is that for the 64th feature, there are actually 2 flags, one flag to indicate that there are more feature flags, and then we extend the record with another `uint64_t` to contain the next 63 feature flags. This should be fine to do, as the flags aren't indicating features of the wallet file itself, but rather capabilities of the last client that opened the wallet. When `x-1` is missing the flags extension flag, version `x` can see that and knows that the capabilities stored in the extension are missing and can therefore do the automatic upgrade.

> Does `VERSION_LATEST` get bumped every time there's a new feature defined, or is that unnecessary?

No, it only needs to be bumped when there is a major incompatible change that cannot be done in the normal wallet feature flags. But for the most part, I don't expect version to ever need to be bumped.

ajtowns reviewed (commented):
Concept ACK 67417e946e1dcf3e5bd8b22be06ad9bba3859efe

Except for lack of clarity on what happens after 64 features, this looks good to me. Should it be added the the 32.0 milestone? Not really clear to me what the priority PRs are for wallet.

ajtowns commented on src/wallet/walletutil.h:
```diff
@@ -56,6 +56,29 @@ enum WalletFlags : uint64_t {
+// than new version numbers.
+// New version numbers must be greater than 299900 which is guaranteed by setting bit 30
+enum WalletClientVersion : int32_t {
+    // OR with all versions to set bit 30
```
Comment needs bumping

achow101 replied:
Ended up dropping and updating the larger comment.

ajtowns commented on src/wallet/wallet.h:
```diff
@@ -436,6 +436,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     std::unordered_map<COutPoint, WalletTXO, SaltedOutpointHasher> m_txos GUARDED_BY(cs_wallet);
 
+    //! Features of the last client to decrypt this wallet
+    std::optional<uint64_t> m_last_decrypted_features;
```
`GUARDED_BY(cs_wallet)` ? if so, `SetLastDecryptedFeatrues()` needs `EXCLUSIVE_LOCKS_REQUIRED`

achow101 replied:
Sure, done.

ajtowns commented on src/wallet/wallet.cpp:
```diff
@@ -3110,7 +3120,7 @@ std::shared_ptr<CWallet> CWallet::CreateNew(WalletContext& context, const std::s
 
     // Initialize version key.
-    if(!WalletBatch(walletInstance->GetDatabase()).WriteVersion(CLIENT_VERSION)) {
+    if(!WalletBatch(walletInstance->GetDatabase()).WriteLastOpenedVersion()) {
```
Should this also initalize the features?

achow101 replied:
Indeed, done

Eunovo commented on src/wallet/walletutil.h:
```diff
@@ -56,6 +56,17 @@ enum WalletFlags : uint64_t {
+// than new version numbers.
+// New version numbers must be greater than 299900 which is guaranteed by setting bit 19
+enum WalletClientVersion : int32_t {
+    MIN_VERSION = (1L << 19),
```
https://github.com/bitcoin/bitcoin/pull/32895/commits/e40b36da5939052ed9e00bde7ee351e1b6193712:

It seems the PR description erroneously mentions setting bit 30.

achow101 replied:
Updated

Eunovo commented on src/wallet/walletdb.cpp:
```diff
@@ -1185,6 +1182,14 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         pwallet->mapMasterKeys.clear();
     }
 
+
```
https://github.com/bitcoin/bitcoin/pull/32895/commits/e40b36da5939052ed9e00bde7ee351e1b6193712:

nit: double blank lines; can fix if retouching

Eunovo commented on src/wallet/walletutil.h:
```diff
@@ -56,6 +56,17 @@ enum WalletFlags : uint64_t {
+// than new version numbers.
+// New version numbers must be greater than 299900 which is guaranteed by setting bit 19
+enum WalletClientVersion : int32_t {
+    MIN_VERSION = (1L << 19),
```
https://github.com/bitcoin/bitcoin/pull/32895/commits/e40b36da5939052ed9e00bde7ee351e1b6193712:

`MIN_VERSION` is not an actual writable version, so it's a bit misleading for it to be part of the enum; consider making it a separate `const`.

achow101 replied:
It is written in the commit that introduces it. I don't think it makes sense to have it be separate, it's defining the lower bound of this enum, just as `VERSION_LATEST` defines the upper bound. Even with future versions, we would never write earlier version numbers, but those versions should be in the enum.

Eunovo reviewed (commented):
Concept ACK https://github.com/bitcoin/bitcoin/pull/32895/commits/3e4490241f26a6932d1d15312d9cb10b35027d48

ajtowns commented:
ACK 3e4490241f26a6932d1d15312d9cb10b35027d48 ; didn't find any problems

Bortlesboat reviewed (commented):
ACK 3e4490241f26a6932d1d15312d9cb10b35027d48

Reviewed the range-diff from 67417e946e1dcf3e5bd8b22be06ad9bba3859efe. The revised wallet-client version namespace, creation-time metadata initialization, and cs_wallet annotations address the intervening review feedback; I did not find additional issues. CI is green.

sedited commented:
@w0xlt can you take another look here?

pablomartin4btc commented on src/wallet/wallet.cpp:
```diff
@@ -3122,9 +3132,16 @@ std::shared_ptr<CWallet> CWallet::CreateNew(WalletContext& context, const std::s
+        }
+        if(!batch.WriteLastOpenedFeatures()) {
+            error = strprintf(_("Error creating %s: Could not write version metadata."), walletFile);
+            return nullptr;
```
in 74b247511c, identical error text ("Could not write version metadata.") for two different failure branches in `CreateNew()` (version-write vs. features-write), making it impossible to tell which one actually failed from the message alone.

Eunovo reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32895/commits/7369187be35e6962c8c1ef123d77e0582a0ea1b2

Bortlesboat reviewed (commented):
ACK 7369187be35e6962c8c1ef123d77e0582a0ea1b2

I reviewed the range-diff from 3e4490241f26a6932d1d15312d9cb10b35027d48; it is just the rebase preserving the upstream `WriteVersion(int)` documentation/argument handling and best-block-locator comment, with no change to the feature-record logic. Built in WSL2 Ubuntu 24.04 and ran `wallet_tests`, `walletdb_tests`, `walletload_tests`, `wallet_crypto_tests`, `init_tests`, plus `wallet_createwallet.py` (RPC and `--usecli`), `wallet_encryption.py`, and `wallet_migration.py` against v28.2; all passed.
